src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -1683,7 +1763,7 @@ export class SavedObjectsRepository {
     if (this._registry.isSingleNamespace(type)) {
       savedObject.namespaces = [SavedObjectsUtils.namespaceIdToString(namespace)];
     }
-    return omit(savedObject, 'namespace') as SavedObject<T>;
+    return omit(savedObject, ['namespace', 'referencesMigrationVersion']) as SavedObject<T>;
```

Comment:
Why are we omitting `referencesMigrationVersion` if we are keeping `migrationVersion`. (But TIL, I wasn't aware `migrationVersion` was exposed...)

Reply:
I just figured there's no reason to expose it to consumers (and doing so would mean I have to change a ton of integration tests).

Reply:
We expose the migrationVersion so that we know which migrations to run (if any) on incoming data. That way if an integration read a document from v7.9.0 and then at a later point writes it back to v7.10.0 we will migrate the v7.9.0 document before writing it into ES.

We want the same behaviour for referencesMigrationVersion, otherwise an integration which does an POST /api/saved_objects/<type>/<id>?overwrite=true will create a document which has invalid references.

Reply:
Core could inject it's own migrations into `buildActiveMigrations` https://github.com/elastic/kibana/blob/master/src/core/server/saved_objects/migrations/core/document_migrator.ts#L194-L222

So that if a dashboard defined a migration for v7.9.0, but core wants to migrate it in v7.11.0 (to perform the referencesMigration) then the `latestVersion` for the dashboard type becomes 7.11.0. The advantage is that the API only has a single `migrationVersion` field. Consumers really shouldn't have to know all the versions of all the kinds of migrations we applied. (And based on #70815 we'll move from `migrationVersion: {dashboard: '7.11.0'}` to `migrationVersion: '7.11.0'`).

The disadvantage is it's less explicit where a migration comes from. If a developer inspects a dashboard they might be surprised that it was migrated in 7.11.0 because they never defined such a migration for their type.

I'm a little bit torn between the developer experience and keeping the API simpler and cleaner. It would be good to hear others thoughts on this?

If we do keep an additional field I would vote for it to be renamed to `coreMigrationVersion` so that it could be used anytime core wants to apply a migration like for instance to implement #70815.

Reply:
@rudolf I'm going to reply to your comments out of order 😄 

> If we do keep an additional field I would vote for it to be renamed to `coreMigrationVersion` so that it could be used anytime core wants to apply a migration like for instance to implement #70815.

Agreed, that sounds better and I'll refer to it as that from now on.

> We want the same behaviour for referencesMigrationVersion, otherwise an integration which does an `POST /api/saved_objects/<type>/<id>?overwrite=true` will create a document which has invalid references.

I respectfully disagree. I feel strongly that `coreMigrationVersion` should be an implementation detail that is not exposed to consumers. I specifically designed this so "convert" and "reference" transforms would not be applied when documents are created, but only during the startup migration process.

We don't enforce referential integrity when objects are created, and we don't know if the outbound reference is valid or not. We do enforce referential integrity (to some extent) when _importing_ objects; in that case, potential ID collisions are detected, IDs are regenerated (in the case of `createNewCopies` mode or "unresolvable conflicts"), and references are updated if need be before any objects are created.

If we silently change reference IDs when objects are created, I think it'll be quite confusing to consumers. Not to mention that to avoid this behavior, consumers would have to specify a `coreMigrationVersion` equal to Kibana's current version when they create objects -- and that input would need to change every time Kibana is upgraded.

> The disadvantage is it's less explicit where a migration comes from. If a developer inspects a dashboard they might be surprised that it was migrated in 7.11.0 because they never defined such a migration for their type.

I don't think we should attempt to expose `coreMigrationVersion` externally or try to tie "reference" transforms to the regular `migrationVersion` instead. These transforms are only applied _because of other objects_ which may potentially be due to an external plugin that does not come bundled with Kibana.

I think it's safe to say:

* When you upgrade Kibana, we guarantee your saved object references will stay intact when object types are converted
* After you upgrade Kibana, your old exports can still be imported, and we guarantee your saved object references will stay intact
* After you upgrade Kibana, if you're manually recreating/updating existing documents using the saved objects APIs, you're on your own to make sure your object IDs are still correct and your references are intact.

Reply:
Renamed `referencesMigrationVersion` to `coreMigrationVersion` in abd7d2b!